scripts/deleteAndLoadSnapsets.js

```diff
@@ -0,0 +1,398 @@
+ * @return {Array} - the values from the gpiiKeyRecords that are snapset GPII Keys.
+ */
+dbLoader.markPrefsSafesGpiiKeysForDeletion = function (gpiiKeyRecords, snapSets) {
+    var gpiiKeysToDelete = [];
+    fluid.each(gpiiKeyRecords.rows, function (gpiiKeyRecord) {
+        var gpiiKey = fluid.find(snapSets, function (aSnapSet) {
+            if (gpiiKeyRecord.value.prefsSafeId === aSnapSet._id) {
+                return gpiiKeyRecord.value;
+            }
+        }, null);
+        if (gpiiKey !== null) {
+            gpiiKey._deleted = true;
+            gpiiKeysToDelete.push(gpiiKey);
+        }
+    });
+    return gpiiKeysToDelete;
+};
+
+/**
+ * Utility to wrap all the pieces to make a bulk documents deletion request
+ * for the snapset Prefs Safes and their associated GPII keys.  Its intended use
+ * is the parameter of the appropriate promise.then() call.
+ * @param {Object} batchDeleteResponse - the reponse handler configured for the
+ *                                       batch delete request.
+ */
+dbLoader.doBatchDelete = function (batchDeleteResponse) {
+    var options = dbLoader.getOptions();
+    var docsToRemove = options.snapsetPrefsSafes.concat(options.gpiiKeys);
+    var execBatchDelete = dbLoader.createBulkDocsRequest(
+        docsToRemove, batchDeleteResponse, options
+    );
+    execBatchDelete();
+};
+
+/**
+ * Create a function that makes a bulk docs POST request using the given data.
+ * @param {Object} dataToPost - JSON data to POST and process in bulk.
+ * @param {Object} responseHandler - http response handler for the request.
+ * @param {Object} options - Data loader options, specifically the POST options.
+ * @return {Function} - A function that wraps an http request to execute the
+ *                      POST.
+ */
+dbLoader.createBulkDocsRequest = function (dataToPost, responseHandler, options) {
+    return function () {
+        var batchPostData = JSON.stringify({"docs": dataToPost});
+        options.postOptions.headers["Content-Length"] = Buffer.byteLength(batchPostData);
+        var batchDocsRequest = http.request(options.postOptions, responseHandler);
+        batchDocsRequest.write(batchPostData);
+        batchDocsRequest.end();
+        return batchDocsRequest;
+    };
+};
+
+/**
+ * Generate a response handler, setting up the given promise to resolve/reject
+ * at the correct time.
+ * @param {Function} handleEnd - Function to call that deals with the response
+ *                               data when the response receives an "end" event.
+ * @param {Promise} promise - Promose to resolve/reject on a response "end" or
+ *                           "error" event.
+ * @param {String} errorMsg - Optional error message to prepend to the error
+ *                            received from a response "error" event.
+ * @return {Function} - Function reponse callback for an http request.
+ */
+dbLoader.createResponseHandler = function (handleEnd, promise, errorMsg) {
```

Comment:
Return a promise rather than accepting one.

Reply:
You've suggested returning a promise in related comments.  I've made changes elsewhere to do that.  But not here, and here's why:  The point of this function was to create and configure a response callback to pass, ultimately, to `http.request()`.  This function, `createResponseHandler()`, returns an anonymous response callback.  Further, that's why the promise was passed in so it could be resolved or rejected based on what happened in the callback.

I tried to rework the logic so as to return a promise instead, but couldn't figure out a graceful way to pass the response handler around.  Also, after the switch to return a promise "everywhere else", and to use `fluid.promise.sequence()` to string the steps together, changing the logic here seemed moot.